logs/querysample/querysample.go

```diff
@@ -0,0 +1,406 @@
+		HasExplain:    true,
+		ExplainSource: pganalyze_collector.QuerySample_AUTO_EXPLAIN_EXPLAIN_SOURCE,
+		ExplainFormat: pganalyze_collector.QuerySample_JSON_EXPLAIN_FORMAT,
+		// Reformat JSON so its the same as when using EXPLAIN (FORMAT JSON)
+		ExplainOutput: "[" + string(explainJSON) + "]",
+	}, nil
+}
+
+func TransformLogExplainJSON(sample state.PostgresQuerySample, normalize bool) (state.PostgresQuerySample, error) {
+	var planDetails []PlanContainer
+
+	if err := json.Unmarshal([]byte(sample.ExplainOutput), &planDetails); err != nil {
+		return sample, err
+	}
+	if len(planDetails) != 1 {
+		return sample, fmt.Errorf("Unexpected plan size: %d (expected 1)", len(planDetails))
+	}
+
+	if normalize {
+		var err error
+		planDetails[0].Plan, err = normalizePlan(planDetails[0].Plan)
+		if err != nil {
+			return sample, err
+		}
+	}
+
+	explainJSON, err := json.Marshal(planDetails)
+	if err != nil {
+		return sample, err
+	}
+
+	sample.ExplainOutput = string(explainJSON)
+
+	return sample, nil
+}
+
+var autoExplainTextPlanDetailsRegexp = regexp.MustCompile(`^Query Text: (.+)\s+([\s\S]+)`)
+
+func transformExplainTextToQuerySample(logLine state.LogLine, explainText string, queryRuntimeMs float64, normalize bool) (state.PostgresQuerySample, error) {
+	if normalize {
+		return state.PostgresQuerySample{}, fmt.Errorf("Requested 'filter_query_sample = normalize', but auto_explain format is text - not supported (discarding sample)")
+	}
+
+	explainParts := autoExplainTextPlanDetailsRegexp.FindStringSubmatch(explainText)
+	if len(explainParts) != 3 {
+		return state.PostgresQuerySample{}, fmt.Errorf("auto_explain output doesn't match expected format")
+	}
+	return state.PostgresQuerySample{
+		Query:         strings.TrimSpace(explainParts[1]),
+		RuntimeMs:     queryRuntimeMs,
+		OccurredAt:    logLine.OccurredAt,
+		Username:      logLine.Username,
+		Database:      logLine.Database,
+		LogLineUUID:   logLine.UUID,
+		HasExplain:    true,
+		ExplainSource: pganalyze_collector.QuerySample_AUTO_EXPLAIN_EXPLAIN_SOURCE,
+		ExplainFormat: pganalyze_collector.QuerySample_TEXT_EXPLAIN_FORMAT,
+		ExplainOutput: explainParts[2],
+	}, nil
+}
+
+func TransformLogMinDurationStatementToQuerySample(logLine state.LogLine, queryText string, queryRuntime string, queryProtocolStep string, parameterParts [][]string) (state.PostgresQuerySample, bool) {
+	// Ignore bind/parse steps of extended query protocol, since they are not the actual execution
+	// See https://www.postgresql.org/docs/current/protocol-flow.html#PROTOCOL-FLOW-EXT-QUERY
+	if queryProtocolStep == "bind" || queryProtocolStep == "parse" {
+		return state.PostgresQuerySample{}, false
+	}
+
+	queryText = strings.TrimSpace(queryText)
+	if queryText == "" {
+		return state.PostgresQuerySample{}, false
+	}
+
+	// Note that we intentionally don't normalize here, because this may be
+	// passed on to log-based EXPLAIN still. We instead normalize this at a
+	// later step (in the output functions).
```

Comment:
Ah, I take it this is why we normalize during explain? Could we instead have a separate `OriginalQuery` field or something in the struct that we use during explain (but not serialize to PB)? It feels brittle to depend on normalization to happen later. Also, does this work if explain is not enabled?

Reply:
This difference (between log-based EXPLAIN and auto_explain normalization) is now gone in the refactoring, as normalization always runs as an explicit step in the logs runner post-processing.

I didn't go with the `OriginalQuery` field idea, since if we wanted to implement this throughout we'd need a few duplicate fields, and I don't think thats necessary since the post-processing is centralized in a function and so its easy to audit that all sensitive fields are taken care of.